internal/storage/download_test.go

```diff
@@ -282,6 +286,109 @@ func TestDownloadToFileError(t *testing.T) {
 
 }
 
+// TestRobustDownloadRetryBadStatus tests file download with retry strategy, when a bad response status is returned
+func TestRobustDownloadRetryBadStatus(t *testing.T) {
+	dir := "_tmp-download"
+	if err := os.MkdirAll(dir, 0755); err != nil {
+		t.Fatalf("failed create temporary directory: %v", err)
+	}
+	// Remove temporary directory at the end
+	defer os.RemoveAll(dir)
+
+	art := &Artifact{
+		FileName: "test.txt", Size: 65536, Link: "http://localhost:43234/test.txt",
+		HashType:  "MD5",
+		HashValue: "ab2ce340d36bbaafe17965a3a2c6ed5b",
+	}
+	// Start Web server
+	srv := Host(":43234", art.FileName, int64(art.Size), false, false, "", "", t)
+	defer srv.Close()
+
+	name := filepath.Join(dir, art.FileName)
+
+	failCountBadStatus = 3
+	if err := downloadArtifact(name, art, nil, "", 1, 0, make(chan struct{})); err == nil {
+		t.Fatal("error expected when downloading artifact, due to bad response status")
+	}
+
+	if err := downloadArtifact(name, art, nil, "", 5, time.Second, make(chan struct{})); err != nil {
+		t.Fatal("expected to handle download error, by using retry download strategy")
+	}
+	check(name, art.Size, t)
+
+	if err := os.Remove(name); err != nil {
+		t.Fatalf("failed to delete test file %s", name)
+	}
+	failCountBadStatus = 2
+	if err := downloadArtifact(name, art, nil, "", 0, 0, make(chan struct{})); err == nil {
+		t.Fatal("error expected when downloading artifact, due to bad response status")
+	}
+}
+
+func TestRobustDownloadRetryCopyError(t *testing.T) {
+	testCopyError(false, false, t)
+	testCopyError(false, true, t)
+	testCopyError(true, false, t)
+	testCopyError(true, true, t)
+}
+
+func testCopyError(withInsufficientRetryCount bool, withCorruptedFile bool, t *testing.T) {
+	dir := "_tmp-download"
+	if err := os.MkdirAll(dir, 0755); err != nil {
+		t.Fatalf("failed create temporary directory: %v", err)
+	}
+	// Remove temporary directory at the end
+	defer os.RemoveAll(dir)
+
+	art := &Artifact{
+		FileName: "test.txt", Size: 65536, Link: "http://localhost:43234/test.txt",
+		HashType:  "MD5",
+		HashValue: "ab2ce340d36bbaafe17965a3a2c6ed5b",
+	}
+	var serverClosing sync.WaitGroup
+	var serverClosed sync.WaitGroup
+	// Start Web server
+	serverClosing.Add(1)
+	serverClosed.Add(1)
+	defer serverClosed.Wait()
+	defer serverClosing.Done()
+	go func() {
+		if withCorruptedFile {
+			corruptFileError = true
+		} else {
+			failCopyError = true
+		}
+		for i := 0; i < 5; i++ {
+			srv := Host(":43234", art.FileName, int64(art.Size), false, false, "", "", t)
+			time.Sleep(2 * time.Second)
+			srv.Close()
+		}
+		failCopyError = false
+		corruptFileError = false
+		srv := Host(":43234", art.FileName, int64(art.Size), false, false, "", "", t)
+		serverClosing.Wait()
+		srv.Close()
+		serverClosed.Done()
+	}()
+
+	name := filepath.Join(dir, art.FileName)
+	retryCount := 10
+	if withInsufficientRetryCount {
+		retryCount = 2
```

Comment:
Why 2 is insufficient retry count? Actually, the download can happen in the first 5 starts for 2 seconds of the test server.

Reply:
Test server is set to fail at least in the first 5 starts(for 10 seconds), so download cannot succeed.
2 tries x 2 seconds in not sufficient to download artifact.
However, 10 tries x 2 seconds must be sufficient.